### PR TITLE
UpsertAssigneeHandler implemented

### PR DESCRIPTION
**Pull request by truhacevkir**

### CODE REVIEW

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/NviService.java [hunk 1]
```diff
@@ -138,10 +140,23 @@ private boolean isNewCandidate(DbCandidate dbCandidate) {
...
-        };
+        if (oldApprovalStatus.status().equals(newApprovalStatus.status())) {
+            return updateAssignee(oldApprovalStatus, newApprovalStatus);
+        } else {
+            return switch (newApprovalStatus.status()) {
+                case APPROVED, REJECTED -> updateStatus(oldApprovalStatus, newApprovalStatus);
+                case PENDING -> resetStatus(oldApprovalStatus);
+            };
+        }
+    }
```
Comment (truhacevkir) [inline nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/NviService.java:151]: In my opinion this object should be call Approval and not ApprovalStatus
And then we can check if update is status, assignee or note and refactor it to do right things

File: nvi-rest/src/main/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandler.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        }
+    }
+
+    private static void hasSameCustomer(ApprovalDto input, RequestInfo requestInfo) throws UnauthorizedException {
+        if (!input.institutionId().equals(requestInfo.getTopLevelOrgCristinId().orElseThrow())) {
+            throw new UnauthorizedException();
+        }
+    }
+
+    private boolean hasAccessRight(User user) {
```
Comment (brinxmat) [inline nvi-rest/src/main/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandler.java:111]: isSomthingToDoWithSpecificRoleInApplication

File: nvi-rest/src/main/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandler.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    private DbApprovalStatus injectAssignee(Candidate candidate, ApprovalDto input) {
+        return getApprovalStatus(candidate, input).copy()
+                   .assignee(new DbUsername(input.assignee()))
+                   .build();
+
+    }
+
+    private static DbApprovalStatus getApprovalStatus(Candidate candidate, ApprovalDto input) {
+        return candidate.approvalStatuses().stream()
+                   .filter(approval -> approval.institutionId().equals(input.institutionId()))
```
Comment (brinxmat) [inline nvi-rest/src/main/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandler.java:87]: Every inline check/filter should be pulled out with a good name.

File: nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        var requestBody = new ApprovalDto(assignee, approvalToUpdate.institutionId());
+        var request = createRequest(randomUUID(), requestBody, assignee);
+        handler.handleRequest(request, output, context);
+        var response = GatewayResponse.fromOutputStream(output, Problem.class);
+
+        assertThat(response.getStatusCode(), is(equalTo(HttpURLConnection.HTTP_NOT_FOUND)));
+    }
+
+    @Test
+    void shouldUpdateAssignee() throws IOException {
```
Comment (brinxmat) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java:104]: When?

File: nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        var response = GatewayResponse.fromOutputStream(output, CandidateResponse.class);
+        var candidateResponse = response.getBodyObject(CandidateResponse.class);
+        var newAssignee = candidateResponse.approvalStatuses().get(0).assignee().value();
+
+        assertThat(newAssignee, is(equalTo(assignee)));
+        assertThat(response.getStatusCode(), is(equalTo(HttpURLConnection.HTTP_OK)));
+    }
+
+    @Test
+    void shouldRemoveAssignee() throws IOException {
```
Comment (brinxmat) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java:121]: When?

File: nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+import no.unit.nva.commons.json.JsonUtils;
+import no.unit.nva.testutils.HandlerRequestBuilder;
+import nva.commons.apigateway.AccessRight;
+import nva.commons.apigateway.GatewayResponse;
+import nva.commons.core.ioutils.IoUtils;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.zalando.problem.Problem;
+
+public class UpsertAssigneeHandlerTest extends LocalDynamoTest {
```
Comment (brinxmat) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java:41]: Not public

File: nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    private Context context;
+    private ByteArrayOutputStream output;
+    private UpsertAssigneeHandler handler;
+    private NviService nviService;
+    private AuthorizedBackendUriRetriever uriRetriever;
+
+    @BeforeEach
+    void init() {
+        output = new ByteArrayOutputStream();
+        context = mock(Context.class);
```
Comment (brinxmat) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java:53]: new FakeContext() ?

File: nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        assertThat(response.getStatusCode(), is(equalTo(HttpURLConnection.HTTP_UNAUTHORIZED)));
+    }
+
+    @Test
+    void shouldReturnUnauthorizedWhenAssigningToUserWithoutAccessRight() throws IOException {
+        mockUserApiResponse("userResponseBodyWithoutAccessRight.json");
+        var candidate = nviService.upsertCandidate(randomApplicableCandidateBuilder()).orElseThrow();
+        var approvalToUpdate = candidate.approvalStatuses().get(0);
+        var assignee = randomString();
+        var requestBody = new ApprovalDto(assignee, approvalToUpdate.institutionId());
```
Comment (brinxmat) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java:81]: Quite a few places
```
    private InputStream createRequest(String assignee) throws JsonProcessingException {
        var candidate = nviService.upsertCandidate(randomApplicableCandidateBuilder()).orElseThrow();
        var approvalToUpdate = candidate.approvalStatuses().get(0);
        var requestBody = new ApprovalDto(assignee, approvalToUpdate.institutionId());
        return createRequest(randomUUID(), requestBody, assignee);
    }
```

File: nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        handler.handleRequest(request, output, context);
+        var response = GatewayResponse.fromOutputStream(output, CandidateResponse.class);
+        var candidateResponse = response.getBodyObject(CandidateResponse.class);
+        var approvalStatus = candidateResponse.approvalStatuses().get(0);
+
+        assertThat(approvalStatus.assignee(), is(nullValue()));
+        assertThat(response.getStatusCode(), is(equalTo(HttpURLConnection.HTTP_OK)));
+    }
+
+    private void persistRandomAssignee(Candidate candidate) {
```
Comment (brinxmat) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/rest/upsert/UpsertAssigneeHandlerTest.java:137]: Is this a good name?

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/NviService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (truhacevkir) [reply nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/NviService.java]: Have refactored following to this:

```
    private DbApprovalStatus toUpdatedApprovalStatus(DbApprovalStatus oldApprovalStatus,
                                                     DbApprovalStatus newApprovalStatus) {
        if (updateIsAssignee(oldApprovalStatus, newApprovalStatus)) {
            return updateAssignee(oldApprovalStatus, newApprovalStatus);
        } else {
            return updateStatus(oldApprovalStatus, newApprovalStatus);
        }
    }

    private static DbApprovalStatus updateStatus(DbApprovalStatus oldApprovalStatus,
                                                        DbApprovalStatus newApprovalStatus) {
        return switch (newApprovalStatus.status()) {
            case APPROVED, REJECTED -> finalizeStatus(oldApprovalStatus, newApprovalStatus);
            case PENDING -> resetStatus(oldApprovalStatus);
        };
    }

    private static boolean updateIsAssignee(DbApprovalStatus oldApprovalStatus, DbApprovalStatus newApprovalStatus) {
        return oldApprovalStatus.status().equals(newApprovalStatus.status());
    }
```

There will be more refactoring when we will add endpoint for updating notes.

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/NviService.java [hunk 10]
```diff
@@ -135,15 +137,36 @@ private boolean isNewCandidate(DbCandidate dbCandidate) {
...
+                                                        DbApprovalStatus newApprovalStatus) {
         return switch (newApprovalStatus.status()) {
-            case APPROVED, REJECTED -> updateStatus(oldApprovalStatus, newApprovalStatus);
+            case APPROVED, REJECTED -> finalizeStatus(oldApprovalStatus, newApprovalStatus);
             case PENDING -> resetStatus(oldApprovalStatus);
         };
     }
 
+    private static boolean updateIsAssignee(DbApprovalStatus oldApprovalStatus, DbApprovalStatus newApprovalStatus) {
+        return oldApprovalStatus.status().equals(newApprovalStatus.status());
```
Comment (hackon) [inline nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/NviService.java:158]: Wouldn't it make more sense to compare assignee to newAssignee?

**Review by hackon — APPROVED** [review]

**Pull request merged**